src/gc/gcinterface.h

```diff
@@ -0,0 +1,396 @@
+    virtual BOOL IsValidSegmentSize(size_t size) = 0;
+    virtual BOOL IsValidGen0MaxSize(size_t size) = 0;
+    virtual HRESULT GarbageCollect(int generation = -1, BOOL low_memory_p = FALSE, int mode = collection_blocking) = 0;
+    virtual void WaitUntilConcurrentGCComplete() = 0;
+    virtual BOOL IsConcurrentGCInProgress() = 0;
+    virtual size_t GetValidSegmentSize(BOOL large_seg = FALSE) = 0;
+    virtual BOOL FinalizeAppDomain(AppDomain *pDomain, BOOL fRunFinalizers) = 0;
+    virtual void SetFinalizeQueueForShutdown(BOOL fHasLock) = 0;
+    virtual size_t GetNumberOfFinalizable() = 0;
+    virtual BOOL ShouldRestartFinalizerWatchDog() = 0;
+    virtual unsigned GetGcCount() = 0;
+    virtual bool IsThreadUsingAllocationContextHeap(gc_alloc_context* acontext, int thread_number) = 0;
+    virtual BOOL IsObjectInFixedHeap(Object *pObj) = 0;
+    virtual BOOL IsEphemeral (Object* object) = 0;
+    virtual uint32_t WaitUntilGCComplete (BOOL bConsiderGCStart = FALSE) = 0;
+    virtual void FixAllocContext (gc_alloc_context* acontext, BOOL lockp, void* arg, void *heap) = 0;
+    virtual int StartNoGCRegion(uint64_t totalSize, BOOL lohSizeKnown, uint64_t lohSize, BOOL disallowFullBlockingGC) = 0;
+    virtual int EndNoGCRegion() = 0;
+    virtual void SetFinalizationRun (Object* obj) = 0;
+    virtual bool RegisterForFinalization (int gen, Object* obj) = 0;
+    virtual size_t GetCurrentObjSize() = 0;
+    virtual void TemporaryEnableConcurrentGC() = 0;
+    virtual void TemporaryDisableConcurrentGC() = 0;
+    virtual BOOL IsConcurrentGCEnabled() = 0;
+    virtual Object* Alloc (gc_alloc_context* acontext, size_t size, uint32_t flags) = 0;
+    virtual Object* Alloc (size_t size, uint32_t flags) = 0;
+    virtual Object* AllocLHeap (size_t size, uint32_t flags) = 0;
+    virtual void PublishObject(uint8_t* obj) = 0;
+    virtual Object* GetContainingObject(void *pInteriorPtr) = 0;
+    virtual void SetGCInProgress(BOOL fInProgress) = 0;
+    virtual CLREventStatic * GetWaitForGCEvent() = 0;
+    virtual void TraceGCSegments() = 0;
+
+#ifdef VERIFY_HEAP
+    virtual void    ValidateObjectMember (Object *obj) = 0;
+    virtual Object * NextObj (Object * object) = 0;
+#endif // VERIFY_HEAP
+
+#ifndef DACCESS_COMPILE    
+    virtual HRESULT WaitUntilConcurrentGCCompleteAsync(int millisecondsTimeout) = 0;    // Use in native threads. TRUE if succeed. FALSE if failed or timeout
+#endif
+
+#if defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)
+    virtual void WalkObject (Object* obj, walk_fn fn, void* context) = 0;
+    virtual void DescrGenerationsToProfiler (gen_walk_fn fn, void *context) = 0;
+#endif //defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)    
+
+    IGCHeap() {}
+    virtual ~IGCHeap() {}
+
+#ifdef STRESS_HEAP
+    //return TRUE if GC actually happens, otherwise FALSE
+    virtual BOOL    StressHeap(gc_alloc_context * acontext = 0) = 0;
+#endif // STRESS_HEAP
+
+    inline static IGCHeap* GetGCHeap()
```

Comment:
I am not happy about the static methods on the interface. I think it should be pure virtual class and the GC heap creation should be done using a global function rather than a static methodr on the interface class. The VM should store the interface pointer after getting it from the construction function. That means that the g_pGCHeap would in the VM as well as the GetGCHeap() function to obtain the pointer.